core_debug.tex

```diff
@@ -130,6 +130,23 @@ \subsection{Icount Trigger}
         very long time to complete.
 \end{steps}
 
+This mechanism cleanly supports a system which supports multiple privilege
+levels, where the OS or a debug stub runs in M-Mode while the program being
+debugged runs in a less privileged mode. When a step is required, the OS or
+debug stub writes \FcsrIcountCount=1, \FcsrIcountAction=0, \FcsrIcountM=0 before
+returning control to the lower user program with an \tt{eret} instruction.
+
+It also allows single step on tiny system which only supports M-Mode, but the
+code becomes tricky to get right. To single step, the debug stub would execute
+something like:
+\begin{verbatim}
+    li    t0, \FcsrIcountCount=4, \FcsrIcountAction=0, \FcsrIcountM=1
+    csrw  tdata1, t0    /* Write the trigger. */
+    lw    t0, 8(sp)     /* Restore t0, count decrements to 3 */
+    lw    sp, 0(sp)     /* Restore sp, count decrements to 2 */
+    ret                 /* Return to program being debugged. count decrements to 1 */
+\end{verbatim}
```

Comment:
I think it should be mret instead of ret.  You got into the single-step handler via a trap which put the return address in mepc instead of x1/ra.  And interrupts are disabled in the handler (which is a good thing because an interrupt coming between the csrw and the return would cause problems) and mret will reenable interrupts.  But otherwise this code seems right.

But I think it's best to not give code examples that are approximately correct and tricky to get right.  Should the spec say "for example, the debug stub can execute this" (giving one possibility for something that definitively works) or should it not provide code at all and just say something more vague about the M-mode code needing to set icount.count to 1 plus the number of instructions after the csr write but before the instruction being stepped?  It's useful to readers to be definitive but more maintainable and less error-prone to be vague.

This is a valuable addition to point out that some implementations may need to consider supporting count>1 (which is not what the spec previously said).

Reply:
I mostly wrote this code to force myself to work through the problem. How about I move it into the appendix?

My main conclusion is that this is probably doable, but getting the code right might be very tricky.